### PR TITLE
Change default initializers for RBM to be scale invariant

### PR DESCRIPTION
**Pull request by PhilipVinc**
This PR does 2 thins: changes the default initialzier of `DenseSymm` and `RBMSymm` to have a variance `1/(features * n_sites)`, and also change the default initializer of `Dense` to `lecun_normal` (so a rescaled normal distrbution with variance `1/n_sites`) instead of the fixed value of `0.01` we had previously. This also affect `RBM`.

With regards to `RBM/Dense`, I think a more appropriate initializer should be variance `1/(alpha*n_sites)`, but flax reccomends `lecun_normal` so i'm not sure what would be best...

before merging this i'd like everyone to run a few tests on this branch, for example for code that uses the default initializers of Dense and DenseSymm.

### CODE REVIEW

**Comment by chrisrothUT** [comment]
I actually think my original instinct was correct that variance ~ 1/(alpha*n_sites^2) for the RBM. 

The idea of variance scaling is that the activations in each layer should stay at variance 1 in each layer. So in order for the activations in the hidden layer of the RBM to have variance 1 we need to set Variance ~ 1/n_sites. 

The issue is that for the RBMSymm, the second layer is a sum layer where we sum over n_symm*features = alpha*n_sites terms. So we need scale the output down by an additional factor of sqrt(alpha*n_sites) or we will start outputting very large numbers. 

So I think the natural solution is to use lecun_normal() for everything and then make sure that we also downscale in sum layers by sqrt(number_of_terms_in_sum).

**Comment by PhilipVinc** [comment]
>  So either we should either set the variance of the weights to be 1/(alpha x n_sites^2) instead of lecun_normal() or we downscale by sqrt(alpha x n_sites) in the sum layer.
I think the latter is a better solution. So we can use lecun_normal() for everything and then make sure that we also downscale in sum layers by sqrt(number_of_terms_in_sum).

I'd like to avoid changing DenseSymm _now_. Such a change would also require you to change something in GCNN i think.

I'd rather just change the initializer to the correct amplitude. Seems `1/(alpha x n_sites^2)`

**Comment by attila-i-szabo** [comment]
> Such a change would also require you to change something in GCNN i think.

Because it's the first layer in a GCNN, it should *require* `lecun_normal` or equivalent to propagate to the next layer correctly. I can't remember if the final layer of the GCNN is normalised as suggested in https://github.com/netket/netket/pull/863#issuecomment-897822235

**Comment by PhilipVinc** [comment]
@chrisrothUT i'm a bit lost.

For `Dense` the input shape is `(N_sites)`  and output is `(alpha x n_sites)`, so to have every `alpha x n_sites` output have the same variance `~1` regardless of system size, we need `lecun_normal` to initialize with variance `1/n_sites`. 
This is good.

The `DenseSymm` layer, instead, has input shape `(N_sites)` and output shape `(features, n_symm)`. [note to self: there's a typo in the docstring].
The symmetrized kernel has shape `(features, n_sites)`, while the full dense kernel that is multiplied by the input has shape `(n_sites, features, n_symm)`.
Since we sum over `n_sites`, if we want the output of every `(feature, n_symm)` to have the same variance `~1` we should also generate the syummetrized kernel with variance `1/n_sites`.

The problem arises becaues we then sum over `n_symm` too, i think..

**Comment by chrisrothUT** [comment]
@PhilipVinc I agree with everything you said in the previous comment. 

In RBMSymm we sum over the output of DenseSymm and the variance of that sum scales like features*n_symm so the final variance is not independent of input size. So we need to scale down the sum layer to make things normalized. 

The other thing we could do is just override the default initializer for DenseSymm when we call RBMSymm

**Comment by chrisrothUT** [comment]
The GCNN overrides the default initializations for DenseSymm and DenseEquivariant at the moment with variance scaling or a user defined initialization. So nothing you do with initialization will mess up the GCNN.

**Comment by PhilipVinc** [comment]
According to this reasoning, shouldn't we also rescale the output of `RBM` (because we are summing over all the `alpha*n_sites` outputs of the dense layer)?

**Comment by chrisrothUT** [comment]
> According to this reasoning, shouldn't we also rescale the output of RBM (because we are summing over all the alpha*n_sites outputs of the dense layer)?

Yes I agree. That's why I think lecun_normal() for everything and scale down sum layers is a good plan.

**Comment by PhilipVinc** [comment]
I'm not sure @gcarleo would agree with replacing all `jnp.sum(x, axis=-1)` with `jnp.sum(x, axis=-1)/x.shape[-1]`...

It's also an extra division (by a constant) to be carried at runtime...

I think a possible alternative would be to wrap the initializer of the kernel into something that divides the initial variance by some value. `alpha*n_sites` for RBM, `features*n_symm` for RBMSymm.

So whatever initializer you provide, it will always be divided by this value (though it's a bit weird...)

**Comment by chrisrothUT** [comment]
> I think a possible alternative would be to wrap the initializer of the kernel into something that divides the initial variance by some value. alpha*n_sites for RBM, features*n_symm for RBMSymm. So whatever initializer you provide, it will always be divided by this value (though it's a bit weird...)

We could also just change the default for `RBMSymm` by calling `DenseSymm(kernel_init= 1/(alpha*n_sites^2))` in `RBMSymm`. Then the user can supply whatever initialization to override the default

**Comment by femtobit** [comment]
Yes, @chrisrothUT's last suggestion sounds good to me as well. We can also add some information to the docstring of `kernel_init` regarding what users need to know w.r.t. scaling when supplying a custom initializer.

**Comment by PhilipVinc** [comment]
I further tweaked the initializer of the RBM (not RBMSymm) and ended up with the following:

 - the new default initializer for RBM kernel is `default_kernel_init = variance_scaling(2, "fan_prod", "truncated_normal")`, which means that the default variance is `2/(N_sites^2*alpha)`. In the plots below you can see that this is a much better initialization for both small (5) and large (30) lattices.
    - I added several docstrings. In particular i mention that the `2` in the variance above comes from the variance of the output of `log_cosh`.
    - This new initializer is the default only for RBM (in general, it should be the default initializer of the last layer of a FFNN with several dense layers).
    - The bias initializer is also a standard trimcated normal with variance `2/(N_sites*alpha)`, in order to keep the correct variance in the output.
    - The visible bias is also changed to be scale invariant: instead of `variance = 0.01` it is now `variance=0.01/N`. Ideally we should initialize it to 0, but then it's much harder to train as there's not always a signal so i noticed that a very small variance works best.

**Comment by PhilipVinc** [comment]
L=5, alpha = 1
download

L=20, alpha=3



L=30, alpha = 1


SGD learning rate is unchanged. This is without SR.

**Comment by gcarleo** [comment]
I don't have strong opinions about this, it seems that the behavior is roughly the same anyways? (maybe marginally better, but  I guess that at large alpha there might more difference)

**Comment by PhilipVinc** [comment]
If you notice with the new initialization there is never an initial plateau in the first few steps. That's the main _added benefit_ of using a proper initialization strategy. 
This is more visible at small system sizes and should reappear at larger size (like N~50 I think) because our old choice (0.01) matches the current for `L~18`.

I don't have a strong opinion either. I was investigating the initializers of DenseSymm but ended up finding that we can improve RBM too...

**Comment by chrisrothUT** [comment]
Yes, one of the pitfalls of bad variance initialization is getting stuck in a ferromagnetic plateau. This is because after you exponentiate you're choosing the phase to be gaussian with the same variance. 

> I further tweaked the initializer of the RBM (not RBMSymm) and ended up with the following:

Why not put the same factor of two in RBMSymm then, since it's related to the nonlinearity? 

> The bias initializer is also a standard trimcated normal with variance 2/(N_sites*alpha), in order to keep the correct variance in the output.

Actually your output variance will be two instead of one with the hidden bias and three with the hidden and visible biases. This still seems fine, you could argue that phases want a variance O(pi). In the GCNN we choose zero-initialization for biases because we have many layers and factors of 2 a can add up.

**Comment by PhilipVinc** [comment]
> Why not put the same factor of two in RBMSymm then, since it's related to the nonlinearity?

I was toying with it, but apparently this initia.lization does not work for `RBMSymm` so well (it works, but you need to lower the learning rate).

> Actually your output variance will be two instead of one with the hidden bias and three with the hidden and visible biases. This still seems fine, you could argue that phases want a variance O(pi). In the GCNN we choose zero-initialization for biases because we have many layers and factors of 2 a can add up.

yes, indeed.
I also read somewhere that biases should ideally be initialzied to zero for deep networks.
But for the RBM case it really seems it's better to initialize them to non zero, that's why i leave them like this.

**Comment by PhilipVinc** [comment]
Curiosly, even using the correct width of `2/alpha*n^2` in RBMSymm leads to very good and stable convergence for L<20 but breaks around L=20... (1D lattice)

**Comment by chrisrothUT** [comment]
> Curiosly, even using the correct width of 2/alpha*n^2 in RBMSymm leads to very good and stable convergence for L<20 but breaks around L=20... (1D lattice)

Does reducing the learning rate fix the problem? It's well known that there's an inverse relationship between model size and learning rate, although I'm not sure that I know of any rigorous quantitative arguments

**Comment by PhilipVinc** [comment]
yes it does, but for example in the case of the standard RBM everything scales roughly the same.
So i thought i could find the same relationship.

**Comment by PhilipVinc** [comment]
@femtobit I rebased this PR to now include only the changes to the initialisers of RBM.
Will modify RBMSymm in a last PR...

If tests pass this should be good to go

**Comment by PhilipVinc** [comment]
Hmm, the failing test

```python
    def test_vmc_functions():
        ha, sx, ma, sampler, driver = _setup_vmc()
        driver.advance(500)

>       assert driver.energy.mean == approx(ma.expect(ha).mean, abs=1e-5)
E       assert DeviceArray(-...dtype=float64) == approx(-5.226...596 ± 1.0e-05)
E         +DeviceArray(-5.22631302, dtype=float64)
E         -approx(-5.226164678266596 ± 1.0e-05)
```

seems to me that it is a bit wrong? we are testing that the energy between two iterations changes by less than 1e-5 also considering the statistical error (and 1e-5 is way too small statistical error)?

**Comment by femtobit** [comment]
> seems to me that it is a bit wrong? we are testing that the energy between two iterations changes by less than 1e-5 also considering the statistical error

Shouldn't this be at the same time step? (I think it's meant to be that way in the test at least. Not sure why we need 500 steps though, that also looks a bit excessive.)

**Comment by PhilipVinc** [comment]
no because `driver.energy` is the last energy computed (which is computed before updating the parameters) and `driver.state.expect` computes it at the current tilmestep, after updating them?

**Comment by femtobit** [comment]
Then the test is probably just wrong (but since energy is usually well converged after 500 steps still passes usually). In any cases, isn't that function just meant to do a full VMC optimization for a few steps to make sure nothing throws an exception etc.?

In that case, maybe we can just change the `assert` to something else that does some reasonable sanity check.

**Comment by PhilipVinc** [comment]
In the end I would rather not merge this, and wait to do a more comprehensive study on the consequences.

While this works very well without using SR, and with SR at sizes `~>=15`, at small sizes it seems that using this initializer it converges to a much worse solution than the previous one.

This test is failing for that reason. because there is still a (albeit small) gradient after 500 iterations with `N=4`.

It might be safe to merge this anyway as I don't think anybody cares for $N<15$.
But to be safe, I'd avoid making a breaking change that we don't understand fully.

**Comment by gcarleo** [comment]
If there is no known optimal way of initializing rbm weights(even though I recall some tricks in Bengio's paper?) I think this should be first explored systematically, potentially in a paper, and then do a PR...
